File: LNCollectionView/LNCollectionView/Layout/LNCollectionViewFlowLayout.m
//
//  LNCollectionViewFlowLayout.m
//  LNCollectionView
//
//  Created by Levison on 16.11.24.
//

#import "LNCollectionViewFlowLayout.h"
#import "LNCollectionView.h"

@interface LNCollectionViewFlowLayout()

@property (nonatomic, assign) CGSize collectionViewContentSize;
@property (nonatomic, copy) NSArray<NSIndexPath *> *indexPathArr;
@property (nonatomic, copy) NSDictionary<NSIndexPath *, NSValue *> *frameDic;
@property (nonatomic, strong) NSMutableDictionary<NSIndexPath *, LNCollectionViewLayoutAttributes *> *attributesMDic;

@end

@implementation LNCollectionViewFlowLayout

- (NSDictionary *)_getSizingInfos
{
    if (self.scrollDirection == LNCollectionViewScrollDirectionHorizontal) {
        CGFloat height = self.collectionView.frame.size.height;
        NSInteger sectionCount = [self.collectionView.dataSource ln_numberOfSectionsInCollectionView:self.collectionView];
        CGFloat cursorY = 0;
        CGFloat cursorX = 0;
        CGFloat lineWidth = 0;
        NSMutableDictionary<NSIndexPath*, NSValue *> *mDic = [[NSMutableDictionary alloc] init];
        NSMutableArray<NSIndexPath *> *indexPathMArr = [[NSMutableArray alloc] init];
        for (int sectionIndex = 0 ; sectionIndex < sectionCount ; sectionIndex ++) {
            NSInteger itemCount = [self.collectionView.dataSource ln_collectionView:self.collectionView numberOfItemsInSection:sectionIndex];
            cursorX = cursorX + self.sectionInset.left;
            lineWidth = 0;
            cursorY = self.sectionInset.top;
            for (int itemIndex = 0; itemIndex < itemCount; itemIndex ++) {
                CGSize itemSize = [(id<LNCollectionViewDelegateFlowLayout>)self.collectionView.delegate ln_collectionView:self.collectionView layout:self sizeForItemAtIndexPath:[NSIndexPath indexPathForItem:itemIndex inSection:sectionIndex]];
                if (cursorY > 0 && cursorY + itemSize.height + self.sectionInset.bottom >= height) {
                    cursorX = cursorX + lineWidth + self.minimumLineSpacing;
                    lineWidth = 0;
                    cursorY = self.sectionInset.top;
                }
                if (lineWidth < itemSize.width) {
                    lineWidth = itemSize.width;
                }
                CGRect itemRect = CGRectMake(cursorX, cursorY, itemSize.width, itemSize.height);
                cursorY = CGRectGetMaxY(itemRect) + self.minimumInteritemSpacing;
                NSValue *value = [NSValue valueWithCGRect:itemRect];
                mDic[[NSIndexPath indexPathForItem:itemIndex inSection:sectionIndex]] = value;
                [indexPathMArr addObject:[NSIndexPath indexPathForItem:itemIndex inSection:sectionIndex]];
            }
            cursorX = cursorX + lineWidth + self.sectionInset.right;
        }
        self.indexPathArr = [NSArray arrayWithArray:indexPathMArr];
        self.collectionViewContentSize = CGSizeMake(cursorX, self.collectionView.bounds.size.height);
        return [NSDictionary dictionaryWithDictionary:mDic];
    } else {
        CGFloat width = self.collectionView.frame.size.width;
        NSInteger sectionCount = [self.collectionView.dataSource ln_numberOfSectionsInCollectionView:self.collectionView];
        CGFloat cursorX = 0;
        CGFloat cursorY = 0;
        CGFloat lineHeight = 0;
        NSMutableDictionary<NSIndexPath *, NSValue *> *mDic = [[NSMutableDictionary alloc] init];
        NSMutableArray<NSIndexPath *> *indexPathMArr = [[NSMutableArray alloc] init];
        for (int sectionIndex = 0 ; sectionIndex < sectionCount ; sectionIndex++) {
            NSInteger itemCount = [self.collectionView.dataSource ln_collectionView:self.collectionView numberOfItemsInSection:sectionIndex];
            cursorY = cursorY + self.sectionInset.top;
            lineHeight = 0;
            cursorX = self.sectionInset.left;
            for (int itemIndex = 0; itemIndex < itemCount ; itemIndex++) {
                CGSize itemSize = [(id<LNCollectionViewDelegateFlowLayout>)self.collectionView.delegate ln_collectionView:self.collectionView layout:self sizeForItemAtIndexPath:[NSIndexPath indexPathForItem:itemIndex inSection:sectionIndex]];
                if (cursorX > 0 && cursorX + itemSize.width + self.sectionInset.right >= width) {
                    cursorY = cursorY + lineHeight + self.minimumLineSpacing;
                    lineHeight = 0;
                    cursorX = self.sectionInset.left;
                }
                if (lineHeight < itemSize.height) {
                    lineHeight = itemSize.height;
                }
                CGRect itemRect = CGRectMake(cursorX, cursorY, itemSize.width, itemSize.height);
                cursorX = CGRectGetMaxX(itemRect) + self.minimumInteritemSpacing;
                NSValue *value = [NSValue valueWithCGRect:itemRect];
                mDic[[NSIndexPath indexPathForItem:itemIndex inSection:sectionIndex]] = value;
                [indexPathMArr addObject:[NSIndexPath indexPathForItem:itemIndex inSection:sectionIndex]];
            }
            cursorY = cursorY + lineHeight + self.sectionInset.bottom;
        }
        self.indexPathArr = [NSArray arrayWithArray:indexPathMArr];
        self.collectionViewContentSize = CGSizeMake(self.collectionView.bounds.size.width, cursorY);
        return [NSDictionary dictionaryWithDictionary:mDic];
    }
}

- (void)prepareLayout
{
    self.frameDic = [self _getSizingInfos];
}

- (void)invalidateLayout
{
    self.indexPathArr = nil;
    self.frameDic = nil;
    [self.attributesMDic removeAllObjects];
    self.collectionViewContentSize = CGSizeZero;
}

- (nullable NSArray<__kindof LNCollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect {
    NSInteger targetIndex = [self binarySearchIndexPathInRect:rect];
    if (targetIndex < 0 || targetIndex >= self.indexPathArr.count) {
        return @[];
    }
    
    NSMutableArray<LNCollectionViewLayoutAttributes *> *result = [NSMutableArray array];
    NSIndexPath *targetIndexPath = self.indexPathArr[targetIndex];
    [result addObject:[self layoutAttributesForItemAtIndexPath:targetIndexPath]];
    
    for (NSInteger i = targetIndex - 1; i >= 0; i--) {
        NSIndexPath *indexPath = self.indexPathArr[i];
        CGRect frame = [self.frameDic[indexPath] CGRectValue];
        if (self.scrollDirection == LNCollectionViewScrollDirectionHorizontal) {
            if (CGRectGetMaxX(frame) < CGRectGetMinX(rect)) {
                break;
            }
        } else {
            if (CGRectGetMaxY(frame) < CGRectGetMinY(rect)) {
                break;
            }
        }
        if (CGRectIntersectsRect(frame, rect)) {
            LNCollectionViewLayoutAttributes *attributes = [self layoutAttributesForItemAtIndexPath:indexPath];
            if (attributes) {
                [result addObject:attributes];
            }
        }
    }

    for (NSUInteger i = targetIndex + 1; i < self.frameDic.allKeys.count; i++) {
        NSIndexPath *indexPath = self.indexPathArr[i];
        CGRect frame = [self.frameDic[indexPath] CGRectValue];
        if (self.scrollDirection == LNCollectionViewScrollDirectionHorizontal) {
            if (CGRectGetMinX(frame) > CGRectGetMaxX(rect)) {
                break;
            }
        } else {
            if (CGRectGetMinY(frame) > CGRectGetMaxY(rect)) {
                break;
            }
        }
        if (CGRectIntersectsRect(frame, rect)) {
            LNCollectionViewLayoutAttributes *attributes = [self layoutAttributesForItemAtIndexPath:indexPath];
            if (attributes) {
                [result addObject:attributes];
            }
        }
    }
    
    return result.copy;
}

- (NSInteger)binarySearchIndexPathInRect:(CGRect)rect {
    if (self.indexPathArr.count == 0) {
        return -1;
    }
    if (self.indexPathArr.count == 1) {
        NSIndexPath *singleKey = self.indexPathArr.firstObject;
        CGRect singleFrame = [self.frameDic[singleKey] CGRectValue];
        if (CGRectIntersectsRect(singleFrame, rect)) {
            return 0;
        } else {
            return -1;
        }
    }
    NSInteger left = 0;
    NSInteger right = self.indexPathArr.count - 1;
    while (left <= right) {
        NSInteger mid = left + (right - left) / 2;
        NSIndexPath *midIndexPath = self.indexPathArr[mid];
        CGRect midFrame = [self.frameDic[midIndexPath] CGRectValue];
        if (CGRectIntersectsRect(midFrame, rect)) {
            return mid;
        }
        if (self.scrollDirection == LNCollectionViewScrollDirectionHorizontal) {
            if (CGRectGetMaxY(midFrame) <= CGRectGetMinY(rect)) {
                left = mid + 1;
            } else if (CGRectGetMinY(midFrame) >= CGRectGetMaxY(rect)) {
                right = mid - 1;
            } else if (CGRectGetMaxX(midFrame) <= CGRectGetMinX(rect)) {
                left = mid + 1;
            } else if (CGRectGetMinX(midFrame) >= CGRectGetMaxX(rect)) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        } else {
            if (CGRectGetMaxX(midFrame) <= CGRectGetMinX(rect)) {
                left = mid + 1;
            } else if (CGRectGetMinX(midFrame) >= CGRectGetMaxX(rect)) {
                right = mid - 1;
            } else if (CGRectGetMaxY(midFrame) <= CGRectGetMinY(rect)) {
                left = mid + 1;
            } else if (CGRectGetMinY(midFrame) >= CGRectGetMaxY(rect)) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
    }
    return -1;
}

- (nullable LNCollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath
{
    if ([self.attributesMDic objectForKey:indexPath]) {
        return [self.attributesMDic objectForKey:indexPath];
    }
    LNCollectionViewLayoutAttributes *att = [LNCollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
    att.frame = [[self.frameDic objectForKey:indexPath] CGRectValue];
    [self.attributesMDic setObject:att forKey:indexPath];
    return att;
    
}

- (CGPoint)targetContentOffsetForProposedContentOffset:(CGPoint)proposedContentOffset withScrollingVelocity:(CGPoint)velocity
{
    return CGPointZero;
}

- (CGPoint)targetContentOffsetForProposedContentOffset:(CGPoint)proposedContentOffset;
{
    return CGPointZero;
}

- (NSMutableDictionary<NSIndexPath *,LNCollectionViewLayoutAttributes *> *)attributesMDic
{
    if (!_attributesMDic) {
        _attributesMDic = [[NSMutableDictionary alloc] init];
    }
    return _attributesMDic;
}

@end

File: Android/app/src/main/java/com/ln/collectionview/collectionview/layout/LNCollectionViewFlowLayout.kt
package com.ln.collectionview.collectionview.layout

import android.graphics.Rect
import android.graphics.RectF
import android.util.Size
import android.util.SizeF
import com.ln.collectionview.collectionview.LNCollectionView
import com.ln.collectionview.collectionview.attributes.LNCollectionViewLayoutAttributes
import kotlin.math.max

interface LNCollectionViewDelegateFlowLayout {

    fun collectionViewSizeForItemAtIndexPath(
        collectionView: LNCollectionView,
        layout: LNCollectionViewLayout,
        indexPath: LNIndexPath
    ): Size {
        return Size(100, 100)
    }

    fun collectionViewMinimumLineSpacingForSectionAtIndex(
        collectionView: LNCollectionView,
        layout: LNCollectionViewLayout,
        section: UInt
    ): Float {
        return 16f
    }

    fun collectionViewMinimumInterItemSpacingForSectionAtIndex(
        collectionView: LNCollectionView,
        layout: LNCollectionViewLayout,
        section: UInt
    ): Float {
        return 8f
    }

    fun collectionViewInsetForSectionAtIndex(
        collectionView: LNCollectionView,
        collectionViewLayout: LNCollectionViewLayout,
        section: UInt
    ): RectF {
        return RectF(0f, 0f, 0f, 0f)
    }

}

class LNCollectionViewFlowLayout(): LNCollectionViewLayout() {

    var indexPathArr: List<LNIndexPath> = listOf()
    var frameDic: Map<String, Rect> = mapOf()
    var attributesMDic: MutableMap<String, LNCollectionViewLayoutAttributes> = mutableMapOf()
    var scrollDirection: LNCollectionViewScrollDirection = LNCollectionViewScrollDirection.VERTICAL
    var defaultLineSpacing: Float = 16f
    var defaultItemSpacing: Float = 8f
    var defaultSectionInset: RectF = RectF(0f,0f,0f,0f)
    private fun getSizingInfos(): Map<String, Rect> {
        val mDic = mutableMapOf<String, Rect>()
        val indexPathMArr = mutableListOf<LNIndexPath>()
        var cursorX = 0f
        var cursorY = 0f
        var lineWidth = 0f
        if (collectionViewWeakRef?.get() == null || collectionViewWeakRef?.get()?.dataSource?.get() == null) {
            return mapOf()
        }
        val validCollectionView = collectionViewWeakRef?.get()!!
        val validDataSource = collectionViewWeakRef?.get()?.dataSource?.get()!!
        val sectionCount = validDataSource.numberOfSectionsInCollectionView(validCollectionView)
        val isHorizontal = scrollDirection == LNCollectionViewScrollDirection.HORIZONTAL
        val frameWidth = validCollectionView.width
        val frameHeight = validCollectionView.height
        if (isHorizontal) {
            for (sectionIndex in 0u until sectionCount) {
                val itemCount = validDataSource.collectionViewNumberOfItemsInSection(validCollectionView, sectionIndex)
                val minLineSpacing = (validCollectionView.delegate?.get() as? LNCollectionViewDelegateFlowLayout)
                    ?.collectionViewMinimumLineSpacingForSectionAtIndex(validCollectionView, this, sectionIndex)
                    ?: defaultLineSpacing
                val minItemSpacing = (validCollectionView.delegate?.get() as? LNCollectionViewDelegateFlowLayout)
                    ?.collectionViewMinimumInterItemSpacingForSectionAtIndex(validCollectionView, this, sectionIndex)
                    ?: defaultItemSpacing
                val sectionInset = (validCollectionView.delegate?.get() as? LNCollectionViewDelegateFlowLayout)
                    ?.collectionViewInsetForSectionAtIndex(validCollectionView, this, sectionIndex)
                    ?: defaultSectionInset
                cursorX += sectionInset.left
                lineWidth = 0f
                cursorY = sectionInset.top
                for (itemIndex in 0u until itemCount) {
                    val itemSize: Size = (validCollectionView.delegate?.get() as? LNCollectionViewDelegateFlowLayout)
                        ?.collectionViewSizeForItemAtIndexPath(validCollectionView, this, LNIndexPath(itemIndex, sectionIndex))
                        ?: defaultItemSize

                    if (cursorY > 0 && cursorY + itemSize.height + sectionInset.bottom >= frameHeight) {
                        cursorX += lineWidth + minLineSpacing
                        lineWidth = 0f
                        cursorY = sectionInset.top
                    }
                    lineWidth = max(lineWidth, itemSize.width.toFloat())
                    val itemRect = Rect(cursorX.toInt(), cursorY.toInt(), (cursorX + itemSize.width).toInt(), (cursorY + itemSize.height).toInt())
                    cursorY = itemRect.bottom + minItemSpacing
                    mDic[LNIndexPath(itemIndex, sectionIndex).key()] = itemRect
                    indexPathMArr.add(LNIndexPath(itemIndex, sectionIndex))
                }
                cursorX += lineWidth + sectionInset.right
            }
        } else {
            cursorX = 0f
            cursorY = 0f
            lineWidth = 0f

            for (sectionIndex in 0u until sectionCount) {
                val itemCount = validDataSource.collectionViewNumberOfItemsInSection(validCollectionView, sectionIndex)
                val minLineSpacing = (validCollectionView.delegate?.get() as? LNCollectionViewDelegateFlowLayout)
                    ?.collectionViewMinimumLineSpacingForSectionAtIndex(validCollectionView, this, sectionIndex)
                    ?: defaultLineSpacing
                val minItemSpacing = (validCollectionView.delegate?.get() as? LNCollectionViewDelegateFlowLayout)
                    ?.collectionViewMinimumInterItemSpacingForSectionAtIndex(validCollectionView, this, sectionIndex)
                    ?: defaultItemSpacing
                val sectionInset = (validCollectionView.delegate?.get() as? LNCollectionViewDelegateFlowLayout)
                    ?.collectionViewInsetForSectionAtIndex(validCollectionView, this, sectionIndex)
                    ?: defaultSectionInset
                cursorY += sectionInset.top
                lineWidth = 0f
                cursorX = sectionInset.left
                for (itemIndex in 0u until itemCount) {
                    val itemSize = (validCollectionView.delegate?.get() as? LNCollectionViewDelegateFlowLayout)
                        ?.collectionViewSizeForItemAtIndexPath(validCollectionView, this, LNIndexPath(itemIndex, sectionIndex))
                        ?: defaultItemSize

                    if (cursorX > 0 && cursorX + itemSize.width + sectionInset.right >= frameWidth) {
                        cursorY += lineWidth + minLineSpacing
                        lineWidth = 0f
                        cursorX = sectionInset.left
                    }
                    lineWidth = max(lineWidth, itemSize.height.toFloat())
                    val itemRect = Rect(cursorX.toInt(), cursorY.toInt(), (cursorX + itemSize.width).toInt(), (cursorY + itemSize.height).toInt())
                    cursorX = itemRect.right + minItemSpacing
                    mDic[LNIndexPath(itemIndex, sectionIndex).key()] = itemRect
                    indexPathMArr.add(LNIndexPath(itemIndex, sectionIndex))
                }
                cursorY += lineWidth + sectionInset.bottom
            }
        }

        collectionViewContentSize = Size(cursorX.toInt(), cursorY.toInt())
        indexPathArr = indexPathMArr
        return mDic
    }

    override fun prepareLayout() {
        frameDic = getSizingInfos()
    }

    override fun invalidateLayout() {
        indexPathArr = listOf()
        frameDic = mapOf()
        attributesMDic.clear()
        collectionViewContentSize = Size(0, 0)
    }

    override fun layoutAttributesForElementsInRect(rect: Rect): List<LNCollectionViewLayoutAttributes> {
        val targetIndex = binarySearchIndexPathInRect(rect)
        if (targetIndex < 0 || targetIndex >= indexPathArr.size) {
            return emptyList()
        }

        val result = mutableListOf<LNCollectionViewLayoutAttributes>()
        val targetIndexPath = indexPathArr[targetIndex]
        result.add(layoutAttributesForItemAtIndexPath(targetIndexPath))

        for (i in targetIndex - 1 downTo 0) {
            val indexPath = indexPathArr[i]
            val frame = frameDic[indexPath.key()] ?: continue
            if (scrollDirection == LNCollectionViewScrollDirection.HORIZONTAL) {
                if (frame.right < rect.left) break
            } else {
                if (frame.bottom < rect.top) break
            }
            if (frame.intersects(rect.left, rect.top, rect.right, rect.bottom)) {
                val attributes = layoutAttributesForItemAtIndexPath(indexPath)
                result.add(attributes)
            }
        }

        for (i in targetIndex + 1 until frameDic.keys.size) {
            val indexPath = indexPathArr[i]
            val frame = frameDic[indexPath.key()] ?: continue
            if (scrollDirection == LNCollectionViewScrollDirection.HORIZONTAL) {
                if (frame.left > rect.right) break
            } else {
                if (frame.top > rect.bottom) break
            }
            if (frame.intersects(rect.left, rect.top, rect.right, rect.bottom)) {
                val attributes = layoutAttributesForItemAtIndexPath(indexPath)
                result.add(attributes)
            }
        }

        return result
    }

    private fun binarySearchIndexPathInRect(rect: Rect): Int {
        if (indexPathArr.isEmpty()) return -1
        if (indexPathArr.size == 1) {
            val singleKey = indexPathArr.first()
            val singleFrame = frameDic[singleKey.key()] ?: return -1
            return if (singleFrame.intersects(rect.left, rect.top, rect.right, rect.bottom)) 0 else -1
        }

        var left = 0
        var right = indexPathArr.size - 1
        while (left <= right) {
            val mid = left + (right - left) / 2
            val midIndexPath = indexPathArr[mid]
            val midFrame = frameDic[midIndexPath.key()] ?: return -1
            if (midFrame.intersects(rect.left, rect.top, rect.right, rect.bottom)) return mid

            if (scrollDirection == LNCollectionViewScrollDirection.HORIZONTAL) {
                if (midFrame.bottom <= rect.top) {
                    left = mid + 1
                } else if (midFrame.top >= rect.bottom) {
                    right = mid - 1
                } else if (midFrame.right <= rect.left){
                    left = mid + 1
                } else if (midFrame.left >= rect.right) {
                    right = mid - 1
                } else {
                    left = mid + 1
                }
            } else {
                if (midFrame.right <= rect.left) {
                    left = mid + 1
                } else if (midFrame.left >= rect.right) {
                    right = mid - 1
                } else if (midFrame.bottom <= rect.top){
                    left = mid + 1
                } else if (midFrame.top >= rect.bottom) {
                    right = mid - 1
                } else {
                    left = mid + 1;
                }
            }
        }
        return -1
    }

    override fun layoutAttributesForItemAtIndexPath(indexPath: LNIndexPath): LNCollectionViewLayoutAttributes {
        return attributesMDic.getOrPut(indexPath.key()) {
            val attributes = LNCollectionViewLayoutAttributes.createForCellAt(indexPath)
            attributes.frame = frameDic[indexPath.key()] ?: Rect()
            attributes
        }
    }
}




